File: 프로그래머스/레벨2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.js
const noteConverter = (melody) => {
    return melody.replace(/[A-Z]#/g, (note) => note[0].toLowerCase());
}

const minuteConverter = (timeString) => {
    const [hour, minute] = timeString.split(":").map(Number);
    return 60 * hour + minute;
}

const melodyRepeater = (melody, playTime) => {
    return melody.repeat(Math.ceil(playTime / melody.length)).substr(0, playTime);
}

function solution(m, musicinfos) {
    const convM = noteConverter(m);
    let resultTitle = "(None)";
    let resultTime = 0;
    
    musicinfos.forEach((musicInfo) => {
       const [start, end, title, melody] = musicInfo.split(",");
        const playTime = minuteConverter(end) - minuteConverter(start);
        const convMelody = noteConverter(melody);
        const repeatedMelody = melodyRepeater(convMelody, playTime);
        
        if (repeatedMelody.includes(convM) && playTime > resultTime) {
            resultTitle = title;
            resultTime = playTime;
        }
    });
    
    return resultTitle;
}

File: 프로그래머스/레벨2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.py
def note_converter(melody):
    result = []
    for note in melody:
        if note != "#":
            result.append(note)
        else:
            result[-1] = result[-1].lower()
    return "".join(result)

def minute_converter(time_string):
    hour, minute = time_string.split(":")
    return 60 * int(hour) + int(minute)

def melody_repeater(melody, play_time):
    return (melody * (play_time // len(melody) + 1))[:play_time]

def solution(m, musicinfos):
    conv_m = note_converter(m)
    result_title = "(None)"
    result_time = 0
    
    for musicinfo in musicinfos:
        start, end, title, melody = musicinfo.split(",")
        play_time = minute_converter(end) - minute_converter(start)
        conv_melody = note_converter(melody)
        repeated_melody = melody_repeater(conv_melody, play_time)
        
        if conv_m in repeated_melody and play_time > result_time:
            result_title = title
            result_time = play_time
            
    return result_title